0000-settlement-engine/0000-settlement-engine.md

```diff
@@ -0,0 +1,391 @@
+
+1. **Send payments**: execute outgoing settlements
+2. **Credit payments**: acknowledge incoming settlements
+
+**Peers**, or entities that transact with one another, may operate compatible settlement engines to settle their liabilities. Different implementations may utilize different settlement systems or types of settlements, such as:
+
+- **Ledger transfers:** adjusting balances on an underlying shared ledger
+- **Physical settlement:** delivering an asset to the counterparty
+- **Payment channels:** sending a claim entitling the peer to an escrow
+- Using a custom protocol to mututally agree to discharge a liability
+
+## Usage in Interledger
+
+Connectors using the [Interledger protocol](https://github.com/interledger/rfcs/tree/master/0001-interledger-architecture) clear and fulfill ILP packets with peers, which represent conditional IOUs that affect their mutual accounting relationships.
+
+Connectors may extend some particular peer a limited line of credit (or none, if they're untrusted). As a connector receives incoming ILP PREPARE packets from a peer, forwards them, and returns corresponding ILP FULFILL packets, the peer's liabilities accrue. If the peer's liabilities to the connector exceeds the credit limit assigned to it, the connector may reject and decline to forward additional packets.
+
+In order to continue transacting, the peer must settle their liabilities. In most cases, this is accomplished through sending a payment on a settlement system that both peers have in common. The connector should acknowledge the incoming payment&mdash;irrevocably discharging some liability the peer owed it&mdash;and enabling it to clear and forward subsequent packets on credit.
+
+Settlement engines provide a standardized mechanism for Interledger connectors to coordinate and reconcile these settlements.
+
+### Accounting
+
+An **account** represents a record of transactions between two peers, denominated in some fungible asset. Each account may be comprised of multiple **account balances**, which each represent the net difference between amounts received (credits) and amounts owed (debits) for some subset of the transactions.
+
+Interledger connectors are RECOMMENDED to operate an **accounting system** which keeps a record of two balances for each account:
+
+- **Accounts payable**, the amount owed by the operator to the counterparty for outgoing obligations its counterparty has fulfilled.
+  - Positive amount indicates the operator is indebted to its counterparty (a _liability_ for the operator).
+  - Negative amount indicates the operator has prefunded.
+- **Accounts receivable**, the amount owed to the operator by its counterparty for incoming obligations the operator has fulfilled.
+  - Positive amount indicates its counterparty is indebted to the operator (an _asset_ to the operator).
+  - Negative amount indicates its counterparty has prefunded.
+
```

Comment:
Good question!

I'd highly recommend implementing it with these 2 values vs 1 single balance, but my goal here is less to explain _how to implement it_ and much more _to explain _the best model to understand the accounting_.

Part of the reason for laying it out this way is, _how would an accountant talk about this interaction? What terms would they use?_

Reply:
Good question!

I'd highly recommend implementing it with these 2 values vs 1 single balance, but my goal here is less to explain _how to implement it_ and much more _to explain the best model to understand the accounting_.

Part of the reason for laying it out this way is asking, _how would an accountant talk about this interaction? What terms would they use?_

On an implementation level, there are a few reasons to prefer this approach:
- A single value doesn't differentiate between a peer owing you money for packets you've fulfilled, or you owing the peer money for packets they've fulfilled, so you're losing information that might be useful. As an example, in Switch, logic around "does the user have enough capacity to receive this payment?" would take into account how much credit has already been extended from the receivable balance, which wouldn't be possible if there was a single balance. (Not saying we need to support that use case, but might be helpful as an example where this might be useful).
- Solves issues with prefunding: if it's a single balance, because there's way to differentiate between the two, sometimes money you prefund to a connector would cause the connector to trigger a settlement back to you, which is bad.

Reply:
> "does the user have enough capacity to receive this payment?"

Do we need capacity to receive money?

Also, it might be better to mention some cases that we might be able to utilize "prefunding".

Reply:
> Why do we need 4 descriptions for these?
> I'd highly recommend implementing it with these 2 values vs 1 single balance,

FWIW, both the [Java](https://github.com/sappenin/java-ilpv4-connector/blob/df/184-initial-module/ilpv4-connector-service-impl/src/main/resources/META-INF/scripts/updateBalanceForFulfill.lua#L23) and [Rust](https://github.com/emschwartz/interledger-rs/blob/master/crates/interledger-store-redis/src/store.rs#L54) implementations currently track balances using two different numbers:

* `clearing_balance`: correlates to the  [Accounts Receivable](https://github.com/interledger/rfcs/pull/536/files#diff-990a915317f4e35578edfa70631bd07dR47) above.
* `prepaid_amount`: correlates to [Accounts Payable](https://github.com/interledger/rfcs/pull/536/files#diff-990a915317f4e35578edfa70631bd07dR44) above.